File: Nhóm 17/NhomThieb/Divide.py
import sys

input = sys.stdin.readline

MAXN = int(5e5 + 5)


class Student:
    def __init__(self, x, r):
        self.x = x
        self.r = r
        self.dis = r - x


def comp(a: Student):
    return (a.x, a.r)


n = int(input())
st = []

for _ in range(n):
    x, r = map(int, input().split())
    st.append(Student(x, r))

# sắp xếp
st.sort(key=comp)

MAX = st[-1].r - st[-1].x
ans = 1
stack = [st[-1].r + st[-1].x]

for i in range(n - 2, -1, -1):
    if st[i].r - st[i].x > MAX:
        ans += 1
        while stack:
            u = stack[-1]
            if u <= st[i].r + st[i].x:
                stack.pop()
                ans -= 1
            else:
                break
        stack.append(st[i].r + st[i].x)
    MAX = max(MAX, st[i].r - st[i].x)

print(ans)

File: Nhóm 17/NhomThieb/Divide.cpp
#include<bits/stdc++.h>
using namespace std;
const int maxn = 5e5 + 5;
long MAX, ans;
stack<int> Arr;
long n;   
struct student{
    long x, r, dis;
} st[maxn];
bool comp(student a, student b){
    if(a.x == b.x)  return a.r < b.r;
    return a.x < b.x;
}
signed main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);  cout.tie(NULL);         
    cin >> n;
    for(long i = 1; i <= n; i++){
        cin >> st[i].x; cin >> st[i].r;
        st[i].dis = st[i].r - st[i].x;
    }
    sort(st + 1, st + n + 1, comp);
    MAX = st[n].r - st[n].x;
    ans = 1;
    Arr.push(st[n].r + st[n].x);
    for(long i = n - 1; i >= 1; i--){
        if(st[i].r - st[i].x > MAX){
            ans++;
            while(!Arr.empty()){
                long u = Arr.top();
                if(u <= st[i].r - st[i].x){
                    Arr.pop();
                    ans--;
                }
                else    break;
            }
            Arr.push(st[i].r - st[i].x);
        } 
        MAX = max(MAX, st[i].r - st[i].x);
    }
    cout << ans;
}